src/osd/PG.cc

```diff
@@ -311,7 +311,14 @@ void PG::proc_master_log(
   // See doc/dev/osd_internals/last_epoch_started
   if (oinfo.last_epoch_started > info.last_epoch_started)
     info.last_epoch_started = oinfo.last_epoch_started;
-  info.history.merge(oinfo.history);
+  if (info.history.merge(oinfo.history)) {
+    if (info.history.last_epoch_clean >= info.history.same_interval_since) {
+      dout(20) << __func__ << "clearing past_intervals" << dendl;
+      past_intervals.clear();
+      dirty_big_info = true;
+    }
+  }
+
```

Comment:
in the peering process, in what situation last_epoch_clean will be >= same_interval_since? last_epoch_clean  is the epoch of last finishing recovery, same_interval_since is the epoch at this peering, right?